manager/state/raft/raft.go

```diff
@@ -75,6 +76,37 @@ const (
 	IsFollower
 )
 
+type lockKeyRotator struct {
+	currentKey []byte
+	newKey     []byte
+	mu         sync.RWMutex
+}
+
+func (l *lockKeyRotator) MaybeUpdateKey(newKey []byte) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if !bytes.Equal(l.currentKey, newKey) {
```

Comment:
The rotation should happen as a callback inside the `DoRotation` function, which grabs a write lock.  I think this should prevent the key from actually being updated whilst in the middle of a rotation?  See https://github.com/docker/swarmkit/pull/1598/files#diff-fcccb5a38e5148a94f6ecfa031233c31R261.  I didn't lock the entire `doSnapshot` call, because it only matters when we actually save a snapshot to disk.

So what will happen if you rotate and rotate again is (at least this is what I was going for...):
1.  MaybeUpdateKey grabs a lock.  Currently no other rotation happening, so it updates the key.
1.  The logic that calls `MaybeUpdateKey` sees if there's a snapshot in progress.  Let's say there isn't for now, so it calls `doSnapshot`
1.  Before `doSnapshot` actually saves a snapshot to disk, it checks if it needs to rote, and if so, actually does rotate out the keys by changing the encoder in side the wrapped snapshotter and WAL writer.  It grabs a lock while doing so.
1.  `saveSnapshot` now starts writing stuff to disk.
1.  Let's say another cluster update comes in.  `MaybeUpdateKey` is triggered again.  If it happens in the middle of the encoder rotation being done in `saveSnapshot`, it will wait until that's done.  Then it updates the key.
1.  Because a snapshot is already in progress, nothing happens - no new snapshot is saved.  The key has been rotated, but it's not being used because the encoders haven't been rotated.
1.  When the snapshot is done, the `snapshotInProgress` channel is written to, then set to nil.  Then it checks to see if a rotation needs to be done.  If so, it triggers another `doSnapshot`.
1.  Which calls `saveSnapshot`, which will rotate the key as necessary, etc.

So if you rotate a key over and over, snapshots will be written over and over to disk, but the cached keys in `keyRotator` can change independently of the actual key used for encryption.  The `keyRotator` is more of an indication of whether the key currently used for encryption needs to change.